src/utils.py

```diff
@@ -0,0 +1,122 @@
+
+    # Calculate average length of moves (ignoring '.' and non-"Play" turns)
+    play_turns['move_length'] = play_turns['move'].str.replace('.', '', regex=False).str.len()
+    avg_move_length = play_turns.groupby(['game_id', 'nickname'])['move_length'].mean().reset_index()
+    avg_move_length.rename(columns={'move_length': 'avg_move_length'}, inplace=True)
+
+    # Calculate the percentage of turns that are "Play"
+    total_turns = turns_df.groupby(['game_id', 'nickname']).size().reset_index(name='total_turns')
+    play_turn_counts = play_turns.groupby(['game_id', 'nickname']).size().reset_index(name='play_turns')
+    play_percentage = pd.merge(total_turns, play_turn_counts, on=['game_id', 'nickname'], how='left')
+    play_percentage['play_percentage'] = (play_percentage['play_turns'] / play_percentage['total_turns']) * 100
+
+    # Calculate the number of turns in each game
+    max_turns = turns_df.groupby('game_id')['turn_number'].max().reset_index()
+    max_turns.rename(columns={'turn_number': 'num_turns'}, inplace=True)
+
+    # Merge all features into a single DataFrame
+
+    features = avg_points_per_move.merge(best_move, on=['game_id', 'nickname'], how='left')
+    features = features.merge(worst_move, on=['game_id', 'nickname'], how='left')
+    features = features.merge(avg_move_length, on=['game_id', 'nickname'], how='left')
+    features = features.merge(play_percentage[['game_id', 'nickname', 'play_percentage']],
+                              on=['game_id', 'nickname'], how='left')
+    features = features.merge(max_turns, on='game_id', how='left')
+
+    return features
+
+
+def parse_features_with_excluded(config_path):
+    """
+    puts features from config nicely into feature and excluded lists, used for log
+    :param config_path:
+    :return:
+    """
+    features = []
+    exclude = []
+
+    with open(config_path, "r") as file:
+        lines = file.readlines()
+
+    # Process each line to detect commented-out features
+    for line in lines:
+        stripped_line = line.strip()
+        if stripped_line.startswith("#"):
+            # Add the commented-out feature to exclude list
+            exclude.append(stripped_line[1:].strip())
+        elif stripped_line.startswith("-"):
+            # Add the active feature to features list
+            features.append(stripped_line[1:].strip())
+
+    # Parse YAML to extract actual config as dictionary
+    with open(config_path, "r") as file:
+        config = yaml.safe_load(file)
+
+    return features, exclude, config
+
+
+def write_to_log(log_path, config_path, model_name, smse):
+    """
+    logs current run to log.txt
+    :param log_path:
+    :param config_path:
+    :param model_name:
+    :param smse:
+    :return: saves run to log
+    """
+    features, exclude, config = parse_features_with_excluded(config_path)
+
+    dont_log = ['player_username', 'game_id', 'player_rating']
+    for feature in dont_log:
+        features.remove(feature)
```

Comment:
```suggestion
    features = [f in feature if f not in dont_log]
```

removing elements from a list id usually considered bad practice